oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/grpc/GRPCCallback.java

```diff
@@ -0,0 +1,145 @@
+
+/**
+ * Use SkyWalking alarm grpc API call a remote methods.
+ */
+@Slf4j
+public class GRPCCallback implements AlarmCallback {
+
+    private AlarmRulesWatcher alarmRulesWatcher;
+
+    private GRPCAlarmSetting alarmSetting;
+
+    private AlarmServiceGrpc.AlarmServiceStub alarmServiceStub;
+
+    private GRPCClient grpcClient;
+
+    public GRPCCallback(AlarmRulesWatcher alarmRulesWatcher) {
+        this.alarmRulesWatcher = alarmRulesWatcher;
+        alarmSetting = alarmRulesWatcher.getGrpchookSetting();
+
+        if (alarmSetting != null) {
+            grpcClient = new GRPCClient(alarmSetting.getTargetHost(), alarmSetting.getTargetPort());
+        }
+    }
+
+    @Override
+    public void doAlarm(List<AlarmMessage> alarmMessage) {
+
+        if (alarmSetting == null) {
+            return;
+        }
+
+        // recreate gRPC client if host and port configuration changed.
+        onGRPCAlarmSettingUpdated(alarmRulesWatcher.getGrpchookSetting());
+
+        ExportStatus status = new ExportStatus();
+
+        StreamObserver<org.apache.skywalking.oap.server.core.alarm.grpc.AlarmMessage> streamObserver =
+            alarmServiceStub.withDeadlineAfter(10, TimeUnit.SECONDS).doAlarm(new StreamObserver<Response>() {
+                @Override
+                public void onNext(Response response) {
+                    // ignore empty response
+                }
+
+                @Override
+                public void onError(Throwable throwable) {
+                    status.done();
+                    if (log.isDebugEnabled()) {
+                        log.debug("Send alarm message failed: {}", throwable.getMessage());
+                    }
+                }
+
+                @Override
+                public void onCompleted() {
+                    status.done();
+                    if (log.isDebugEnabled()) {
+                        log.debug("Send alarm message successful.");
+                    }
+                }
+            });
+
+        alarmMessage.forEach(message -> {
+            org.apache.skywalking.oap.server.core.alarm.grpc.AlarmMessage.Builder builder =
+                org.apache.skywalking.oap.server.core.alarm.grpc.AlarmMessage.newBuilder();
+
+            builder.setScopeId(message.getScopeId());
+            builder.setScope(message.getScope());
+            builder.setName(message.getName());
+            builder.setId0(message.getId0());
+            builder.setId1(message.getId1());
+            builder.setRuleName(message.getRuleName());
+            builder.setAlarmMessage(message.getAlarmMessage());
+            builder.setStartTime(message.getStartTime());
+
+            streamObserver.onNext(builder.build());
+        });
+
+        streamObserver.onCompleted();
+
+        long sleepTime = 0;
+        long cycle = 100L;
+        /**
+         * For memory safe of oap, we must wait for the peer confirmation.
+         */
+        while (!status.isDone()) {
```

Comment:
There may be thread-visibility issue,  `org.apache.skywalking.oap.server.core.exporter.ExportStatus#done` is not `volatile`, BTW, `ExportStatus` can be annotated with Lombok annotations

Reply:
Yes, I agree. I missed that last time.

Reply:
> ExportStatus is in exporter, why use it in the alarm?

Maybe a `CountDownLatch` makes it more convenient?

Reply:
It works, or move `ExportStatus` to higher-level package, such as `library-util`.

Reply:
Considering `exporter grpc`(https://github.com/apache/skywalking/blob/5b255ba3e3b50e0e0196b28cdca21dcd5c9c3da2/oap-server/exporter/src/main/java/org/apache/skywalking/oap/server/exporter/provider/grpc/GRPCExporter.java#L173)  have the same problem and for reuse `ExportStatus`, move to higher-level package just ok.